core/types/src/keys.rs

```diff
@@ -0,0 +1,356 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Basic key types
+
+use core::{
+    fmt::{Debug, Display},
+    marker::PhantomData,
+};
+
+use curve25519_dalek::{ristretto::RistrettoPoint, scalar::Scalar};
+use zeroize::Zeroize;
+
+use mc_crypto_keys::{KeyError, RistrettoPrivate, RistrettoPublic};
+
+// Exported key types
+
+/// Subaddress view private key
+pub type SubaddrViewPrivate = Key<Subaddr, View, RistrettoPrivate>;
+/// Subaddress spend private key
+pub type SubaddrSpendPrivate = Key<Subaddr, Spend, RistrettoPrivate>;
+
+/// Subaddress view public key
+pub type SubaddrViewPublic = Key<Subaddr, View, RistrettoPublic>;
+/// Subaddress spend public key
+pub type SubaddrSpendPublic = Key<Subaddr, Spend, RistrettoPublic>;
+
+/// Root view private key
+pub type RootViewPrivate = Key<Root, View, RistrettoPrivate>;
+/// Root spend private key
+pub type RootSpendPrivate = Key<Root, Spend, RistrettoPrivate>;
+
+/// Root view public key
+pub type RootViewPublic = Key<Root, View, RistrettoPublic>;
+/// Root spend public key
+pub type RootSpendPublic = Key<Root, Spend, RistrettoPublic>;
+
+// Markers
+
+/// Subaddress marker type
+#[derive(Copy, Clone, Debug)]
+pub struct Subaddr;
+
+/// Root address marker type
+#[derive(Copy, Clone, Debug)]
+pub struct Root;
+
+/// View key marker type
+#[derive(Copy, Clone, Debug)]
+pub struct View;
+
+/// Spend key marker type
+#[derive(Copy, Clone, Debug)]
+pub struct Spend;
+
+/// Generic key object, see type aliases for use
+#[derive(Clone, Debug, Zeroize)]
+pub struct Key<ADDR, KIND, KEY: Default + Zeroize> {
+    key: KEY,
+    #[zeroize(skip)]
+    _addr: PhantomData<ADDR>,
+    #[zeroize(skip)]
+    _kind: PhantomData<KIND>,
+}
+
+/// Explicit conversion to internal key type for backwards compatibility
+impl<ADDR, KIND, KEY: Default + Zeroize> Key<ADDR, KIND, KEY> {
+    pub fn inner(self) -> KEY {
+        self.key
+    }
+}
+
+/// AsRef to internal key type for backwards compatibility
+impl<ADDR, KIND, KEY: Default + Zeroize> AsRef<KEY> for Key<ADDR, KIND, KEY> {
+    fn as_ref(&self) -> &KEY {
+        &self.key
+    }
+}
+
+/// [`core::fmt::LowerHex`] where `KEY: AsRef<[u8]>`
+impl<ADDR, KIND, KEY: AsRef<[u8]> + Default + Zeroize> core::fmt::LowerHex
+    for Key<ADDR, KIND, KEY>
+{
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        let data = self.key.as_ref();
+        for d in data {
+            write!(f, "{:02x}", d)?;
```

Comment:
I think we have avoided putting out private keys when doing string representations (e.g. https://github.com/mobilecoinfoundation/mobilecoin/blob/master/crypto/keys/src/ristretto.rs#L150)
That reduces the chance of the key ending in a log message somewhere.

Reply:
Maybe this is fine due to https://github.com/mobilecoinfoundation/mobilecoin/blob/328df8f3e51aa461df4232393eeb0b56a9c6d47d/core/types/src/keys.rs#L289